### ISSUE TITLE
Feature proposal for individual partition round-robin queue polling for consumers

### ISSUE THREAD

**Issue by accelerated** [mfontanini/cppkafka#62]
Hi @mfontanini, so in the light of what I discovered in reference to the behavior of the global queue forwarding (see comments [here](https://github.com/edenhill/librdkafka/issues/1395#issuecomment-384058260)), there is definitely a need for round-robin polling on the partitions instead of the default behavior which is to consume one batch from partition A, followed by another batch from partition B, etc. This might be OK with some applications but it's _not_ for my case and it's quite detrimental if not done properly. So the proposal would be to support both types of consume scenarios in your library. 

1) Enabling the feature could be done by:
* Provide some setter in the consumer class which would tell which polling mechanism to use (could turn on/off the feature at will). The setter would call `rd_kafka_poll_set_consumer(NULL)` if it's disabled. 
* Add a new constructor with an additional parameter which would only enable/disable the type of polling once per object. In this case we simply would not call `rd_kafka_poll_set_consumer(handle)`.
* Add a default parameter to the existing constructor which might break ABI (not desired).
* Add a cppkafka-specific configuration option for consumers which could be passed as part of the current constructor `Consumer(Configuration config)`. This could be the least invasive. 

2) Adding a `Queue` wrapper class to manage the queue handle returned by `rd_kafka_queue_get_partition`. The destructor would call the appropriate destroy function in librdkafka. This could be implemented in a similar way to your `Topic::HandlePtr` unique ptr with a destructor template argument.

3) Changing polling method by:
* keep existing `Consumer::poll()` function and call `rd_kafka_consumer_poll()` if partition polling is not enabled. 

    If enabled do:
- call `rd_kafka_poll()` to get general events.
- for each topic partition in the topic partition list (a copy of the topic partition list will be kept during the call to `assign` - see 4) call `rd_kafka_consume_queue()`. Note that only one partition will be consumed on each call to `Consumer::poll()` as a static round-robin counter will be kept internally to the function. 
- return the message if any (_or even better, poll until one valid message is returned or all partitions have been polled_)

  3.a) Alternatively keep `poll()` unchanged and add `poll_partitions()` function however in this case some error must be thrown if the user enables partition polling and then calls `poll()`.

4) Upon rebalance assign, keep a local copy of the `TopicPartionsList`. Upon revoke, clear this list. Perhaps instead of keeping a copy of the TopicPartitionsList, when assign is called, we can at that point create a list of Queue objects belonging to each partition. That would be the simplest way.

**Comment by mfontanini** [comment]
Let's wait until the edenhill replies on that issue or at least you manage to make it work but this sounds like a good idea. I wish rdkafka had better documentation on this, as I actually implemented this sort of round robin consumption and it took me a good amount of effort as I ended up pausing/resuming consumption on different partitions as I consumed them. 

I'm not entirely sure I understand the API you're proposing but I think having a `Queue` class works great but it should be the user's responsibility to poll from each of them. e.g. in my case I didn't want exactly round robin but instead I was doing sorted consumption based on some timestamp stored in the contents of each message. If the polling was opaque and implemented as actual round robin, this wouldn't fix my case. However, if as a user I got a `Queue` class and decided to poll on each of them individually, that would work. I'm not sure if you're proposing exactly this, so sorry if it's actually the same.
